### ISSUE TITLE
Sync warp redownload the state of parachain on restart

### ISSUE THREAD

**Issue by crystalin** [paritytech/substrate#14012]
### Description of bug

When using `--sync warp` on both the parachain and relaychain side (on Moonbeam node using Substrate 0.9.40), when downloading the parachain state, it waits for the relaychain to fully sync all the blocks:
![image](https://user-images.githubusercontent.com/329248/234338545-31c1b6fc-66c8-4f95-886c-d4cfc6f6891e.png)

If the node restarts during that time, it needs to redownload the state :( (which takes ~ 2 hours)
<img width="1252" alt="image" src="https://user-images.githubusercontent.com/329248/234338809-ab83eb6a-80fa-4d0b-add5-df98c2bfaba1.png">

### Steps to reproduce

Using moonbeam branch `https://github.com/PureStake/moonbeam/tree/upgrade-v0.9.40`
```
~/moonbeam-warp --base-path ./networks/moonbeam --chain moonbeam --database paritydb --blocks-pruning 2048 --state-pruning 2048 --sync warp -- --sync warp --state-pruning 256
```

**Comment by arkpar** [comment]
Looking at the log above, parachain warp sync waits for the relay chain to complete full block history download, which is wrong. It should start as soon as the relay chain syncs and catches up the latest finalized state. No need to wait for historical block download to complete.

**Comment by crystalin** [comment]
@arkpar Yes that is my observation too, it also happens on Moonriver (Kusama), so must be a wrong behavior in the code

**Comment by bkchr** [comment]
> Looking at the log above, parachain warp sync waits for the relay chain to complete full block history download, which is wrong. It should start as soon as the relay chain syncs and catches up the latest finalized state. No need to wait for historical block download to complete.

Not sure where you see this. Looking at the logs above you see that the relay chain is downloading the block history while the parachain is already downloading the state. So everything works as expected.

**Comment by crystalin** [comment]
> Not sure where you see this. Looking at the logs above you see that the relay chain is downloading the block history while the parachain is already downloading the state. So everything works as expected.

No, the state is already downloaded, as you can see the size of the state is the same (it was the exact same value for hours)

**Comment by crystalin** [comment]
@bkchr Actually, even when the relaychain history is done, the state is not imported:
<img width="1155" alt="image" src="https://user-images.githubusercontent.com/329248/234389913-3d48a642-7f9b-409e-a466-199a505b74af.png">

**Comment by bkchr** [comment]
What is the CPU usage? 15gb is quite a lot.

**Comment by crystalin** [comment]
CPU is fairly low (not idle however)
<img width="1824" alt="image" src="https://user-images.githubusercontent.com/329248/234390742-471accf1-d534-439e-a93e-2404a998811d.png">

**Comment by crystalin** [comment]
There nothing in the perf analysis neither:
<img width="1742" alt="image" src="https://user-images.githubusercontent.com/329248/234391641-108fa849-cd5a-4bf6-a392-991feb580438.png">

**Comment by crystalin** [comment]
> What is the CPU usage? 15gb is quite a lot.

Yes, Moonbeam is getting big

**Comment by crystalin** [comment]
The downloading ended ~2h ago:
<img width="1167" alt="image" src="https://user-images.githubusercontent.com/329248/234392325-4ebc63b8-a526-4367-8caa-6aa8c210f840.png">

**Comment by bkchr** [comment]
@crystalin can you run with `sync=debug`? Just restart and don't delete the relay chain db.

**Comment by crystalin** [comment]
It will take an hour or 2, I'll upload the logs tomorrow morning

**Comment by bkchr** [comment]
Perfect, ty!

**Comment by arkpar** [comment]
> Not sure where you see this. Looking at the logs above you see that the relay chain is downloading the block history while the parachain is already downloading the state. So everything works as expected.

That's true. I did not notice that the block history download is still in progress. Although it is almost over. Perhaps we should not start it until the parachain warp sync is complete, so it does not compete for traffic. 

> @arkpar afair we store the entire state in memory until we have downloaded everything and then we flush it to the db? This explains why it re-downloads the state after restart.

That's true. Current implementation requires all state to fit in memory and therefore does not really support huge states. We could potentially save the data to a temporary file and then commit it to the database in batches. 

Another issue is that warp sync does not take `--block-pruning` into account and always downloads full history.

**Comment by bkchr** [comment]
> Another issue is that warp sync does not take `--block-pruning` into account and always downloads full history.

Yeah I know. I need to write some issue for this any way. I think that we added block pruning too early and if too many nodes run with this, we would have problems. But yeah, this is a topic for another issue.

**Comment by crystalin** [comment]
The logs are quite big 4GB.
I did run `egrep -v 'Too many full nodes|Failed to register| disconnected' moonbeam.log > moonbeam.filtered.log` to reduce it to 100Mb (we probably should prevent the node from spamming those logs)
You can access it there: https://alan-stuff.s3.amazonaws.com/moonbeam.filtered.log

**Comment by bkchr** [comment]
```
2023-04-25 23:41:59.185 DEBUG tokio-runtime-worker sync: [🌗] Importing state data from 12D3KooWS7JaE5qUmdcg57CcGW4otSa3aq3BabXGRsuX53WuM8Vt with 0 keys, 1311512 proof nodes.    
2023-04-25 23:41:59.185 DEBUG tokio-runtime-worker sync: [🌗] Importing state from 1311512 trie nodes    
2023-04-25 23:41:59.201 DEBUG tokio-runtime-worker sync: [🌗] Imported with 18459 keys    
2023-04-25 23:41:59.516 DEBUG tokio-runtime-worker sync: [🌗] Importing state data from 12D3KooWS7JaE5qUmdcg57CcGW4otSa3aq3BabXGRsuX53WuM8Vt with 0 keys, 727313 proof nodes.    
2023-04-25 23:41:59.516 DEBUG tokio-runtime-worker sync: [🌗] Importing state from 727313 trie nodes    
2023-04-25 23:41:59.524 DEBUG tokio-runtime-worker sync: [🌗] Imported with 7739 keys    
2023-04-25 23:41:59.525 DEBUG tokio-runtime-worker sync: [🌗] State download is complete. Import is queued  
```

It finishes the download as I had assumed and then also queues it.

**Comment by bkchr** [comment]
Yeah for sure :P Now we need to find out why :D

**Comment by crystalin** [comment]
let me know if you want me to try more... I can rely on Moonriver node which is "only" 5Gb

**Comment by bkchr** [comment]
Reporting back after the next run.

**Comment by bkchr** [comment]
Okay, took longer than I had expected :see_no_evil: 

So, the things you reported in this issue are actually two different things.

1. The re-download, as explained above we only store the state in memory and thus, it needs to be re-downloaded after a restart. We should probably create a new issue for this. However, it will not be that easy as you could maybe started downloading the state for block 100 and next time you restarted your node the chain is already at block 400 and only archive nodes have the state of block 100. So, what are you going to do? But yeah, for restarting a node in between it makes sense to cache it.

2. The second issue is actually a bug in the nimbus verifier. You are calling into the state to check the inherents. However, this doesn't work with state sync as you first need to import the state, which goes through the same pipeline. So, the fix is the following patch:
```
diff --git a/nimbus-consensus/src/import_queue.rs b/nimbus-consensus/src/import_queue.rs
index 4641f09..29600ef 100644
--- a/nimbus-consensus/src/import_queue.rs
+++ b/nimbus-consensus/src/import_queue.rs
@@ -56,6 +56,20 @@ where
                &mut self,
                mut block_params: BlockImportParams<Block, ()>,
        ) -> Result<BlockImportParams<Block, ()>, String> {
+               // Skip checks that include execution, if being told so or when importing only state.
+               //
+               // This is done for example when gap syncing and it is expected that the block after the gap
+               // was checked/chosen properly, e.g. by warp syncing to this block using a finality proof.
+               // Or when we are importing state only and can not verify the seal.
+               if block_params.with_state() || block_params.state_action.skip_execution_checks() {
+                       // When we are importing only the state of a block, it will be the best block.
+                       block_params.fork_choice = Some(sc_consensus::ForkChoiceStrategy::Custom(
+                               block_params.with_state(),
+                       ));
+
+                       return Ok(block_params);
+               }
+
                debug!(
                        target: crate::LOG_TARGET,
                        "🪲 Header hash before popping digest {:?}",
```

And the following pr would have made this entire thing here much easier, as we would have seen the error directly: https://github.com/paritytech/substrate/pull/14025
